rust/src/ldap/logger.rs

```diff
@@ -0,0 +1,190 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Giuseppe Longo <giuseppe@glongo.it>
+
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+use crate::ldap::filters::*;
+use crate::ldap::ldap::LdapTransaction;
+use crate::ldap::types::*;
+
+fn log_ldap(tx: &LdapTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("ldap")?;
+
+    if let Some(req) = &tx.request {
+        js.open_object("request")?;
+        js.set_uint("message_id", req.message_id.0 as u64)?;
+
+        match &req.protocol_op {
+            ProtocolOp::SearchRequest(msg) => log_search_request(msg, js)?,
+            ProtocolOp::BindRequest(msg) => log_bind_request(msg, js)?,
+            ProtocolOp::UnbindRequest => log_unbind_request(js)?,
+            _ => {}
+        };
+
+        log_controls(&req.controls, js)?;
+
+        js.close()?;
+    }
+
+    if tx.responses.len() > 0 {
+        js.open_array("responses")?;
+
+        for response in &tx.responses {
+            js.start_object()?;
+            js.set_uint("message_id", response.message_id.0 as u64)?;
+            match &response.protocol_op {
+                ProtocolOp::SearchResultEntry(msg) => log_search_result_entry(msg, js)?,
+                ProtocolOp::SearchResultDone(msg) => log_search_result_done(msg, js)?,
+                ProtocolOp::BindResponse(msg) => log_bind_response(msg, js)?,
+                _ => {}
+            }
+            log_controls(&response.controls, js)?;
+            js.close()?;
+        }
+        js.close()?;
+    }
+
+    js.close()?;
+    Ok(())
+}
+
+fn log_search_request(msg: &SearchRequest, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.set_string("operation", "search_request")?;
```

Comment:
Guess it would be nice if there would be one unique `js.set_string("operation", protocol_op.to_str())` or something similar...